phastSim/phastSim.py

```diff
@@ -641,23 +809,196 @@ def sampleMutation(self, allele, rates, rand):
+            # nodes need to be created if on a different level, otherwise can directly change nodes
+            child0, child1 = node.belowNodes[0], node.belowNodes[1]
+            rate0, rate1 = child0.rate, child1.rate
+            newRate0, newRate1 = rate0, rate1
+            newChild0, newChild1 = child0, child1
+
+            # deal with the left child
+            deleted_leaves_left = 0
+            if rand > rate0 or remaining_deletions == 0:
+                # under these conditions we can short circuit and ignore this whole branch
+                pass
+
+            else:
+                if child0.level < level or not (child0.isTerminal and child0.allele == "-"):
+                    newChild0 = genomeNode(level=level)
+                    newChild0.isTerminal = child0.isTerminal
+                    newChild0.rate = child0.rate
+
+                    if child0.isTerminal:
+                        newChild0.refNode = child0.refNode
+                        newChild0.allele = child0.allele
+                    else:
+                        newChild0.belowNodes = list(child0.belowNodes)
+
+                
+                deleted_leaves_left = self.deleteNodes(rand, newChild0, remaining_deletions, level)
+                newRate0 = newChild0.rate
+
+            # deal with right child but with slightly different exit conditions
+            deleted_leaves_right = 0
+            if rand > rate0 + rate1 or remaining_deletions - deleted_leaves_left == 0:
+                # can short circuit this child under slightly different conditions
+                pass
+
+            else:
+                if child1.level < level or not (child1.isTerminal and child1.allele == "-"):
+                    newChild1 = genomeNode(level=level)
+                    newChild1.isTerminal = child1.isTerminal
+                    newChild1.rate = child1.rate
+
+                    if child1.isTerminal:
+                        newChild1.refNode = child1.refNode
+                        newChild1.allele = child1.allele
+
+                    else:
+                        newChild1.belowNodes = list(child1.belowNodes)
+       
+
+                deleted_leaves_right = self.deleteNodes(rand - rate0, newChild1, remaining_deletions - deleted_leaves_left, level)
+                newRate1 = newChild1.rate
+
+            node.rate = newRate0 + newRate1
+            node.belowNodes = [newChild0, newChild1]
+
+            return deleted_leaves_left + deleted_leaves_right
+
+    def buildTree(self, node, insert_list, pos_left, pos_right):
+
+        if pos_left == pos_right:
+            # then we are at a leaf, we need to build a new leaf node from scratch
+            node.isTerminal = True
+            node.refNode = node
+            if self.codon:
+                node.allele = self.codonAlleles[insert_list[0]]
+            else:
+                
+                node.allele = self.alleles[insert_list[0]]
+                node.rates = self.mutMatrixRepeats[-1]
```

Comment:
This line would need to be replaced, with the same process that is used to sample mutation rates for the terminal nodes of the original genome tree. This means considering hypermutability and gamma variation in rates. In my original code this was done in function populateGenomeTree(), lines 476-564. Not sue if it would be cleaner to rewrite those lines as its own function, and then call that function here.

Reply:
Also note that this line does not create a copy of self.mutMatrixRepeats[-1], so when you modify it, you also modify the rates for all other inserted nodes. This is why below I suggest to create a new array similar to self.mutMatrixRepeats but specific for this particular insertion event.